frames/src/test/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModelTest.kt

```diff
@@ -0,0 +1,231 @@
+import com.checkout.frames.style.component.base.InputFieldStyle
+import com.checkout.frames.style.view.InputFieldViewStyle
+import com.checkout.frames.view.InputFieldState
+import com.checkout.validation.api.CVVComponentValidator
+import com.checkout.validation.model.ValidationResult
+import io.mockk.MockKAnnotations
+import io.mockk.every
+import io.mockk.impl.annotations.RelaxedMockK
+import io.mockk.impl.annotations.SpyK
+import io.mockk.junit5.MockKExtension
+import io.mockk.verify
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.StandardTestDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.amshove.kluent.internal.assertEquals
+import org.amshove.kluent.internal.assertFalse
+import org.junit.After
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+
+@ExperimentalCoroutinesApi
+@SuppressLint("NewApi")
+@ExtendWith(MockKExtension::class)
+internal class CVVInputFieldViewModelTest {
+
+    @RelaxedMockK
+    private lateinit var mockCVVComponentValidator: CVVComponentValidator
+
+    @SpyK
+    private lateinit var spyInputFieldStyleMapper: Mapper<InputFieldStyle, InputFieldViewStyle>
+
+    @SpyK
+    private lateinit var spyInputFieldStateMapper: Mapper<InputFieldStyle, InputFieldState>
+
+    private lateinit var viewModel: CVVInputFieldViewModel
+
+    private val dispatcher = StandardTestDispatcher()
+
+    @RelaxedMockK
+    private lateinit var cvvComponentConfig: CVVComponentConfig
+
+    init {
+        initMappers()
+        initConfiguration()
+    }
+
+    @BeforeEach
+    fun setUp() {
+        MockKAnnotations.init(this)
+        Dispatchers.setMain(dispatcher)
+        every { cvvComponentConfig.cardScheme } returns CardScheme.VISA
+        every { cvvComponentConfig.cvvInputFieldStyle } returns InputFieldStyle()
+
+        viewModel = CVVInputFieldViewModel(
+            spyInputFieldStateMapper, spyInputFieldStyleMapper, cvvComponentConfig, mockCVVComponentValidator
+        )
+    }
+
+    @After
+    fun after() {
+        Dispatchers.resetMain()
+    }
+
+    /** Initial state tests **/
+
+    @Test
+    fun `when view model is initialised then CVVInputField style is mapped to initial CVVInputField state`() {
+        // Then
+        verify { spyInputFieldStateMapper.map(any()) }
+    }
+
+    @Test
+    fun `when view model is initialised then initial state has empty cvv text`() {
+        // Then
+        assertTrue(viewModel.cvvInputFieldState.cvv.value.isEmpty())
+    }
+
+    @Test
+    fun `when view model is initialised then initial state has max cvv length updated as per card scheme from the cvvComponentConfig`() {
+        // Then
+        assertEquals(cvvComponentConfig.cardScheme.cvvLength.max(), viewModel.cvvInputFieldState.cvvLength.value)
```

Comment:
shall we add a test to cover that the `cvvLength` is actually taking effect?

Reply:
This test case covers cvvLength and is updated accordingly for the card scheme. However, it is a valid test to improve and it's part of the input field. We have not added any test cases for compose inputfield yet. We can add it to test improvements except if you think we can add validation for this in the viewmodel.